### PR TITLE
MarkdownBear: Add `validate-links` plugin

### PR DESCRIPTION
**Pull request by yash-nisar**
Closes https://github.com/coala/coala-bears/issues/924

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

**Comment by yash-nisar** [comment]
CC @Makman2

File: bears/markdown/MarkdownBear.py [hunk 1]
```diff
@@ -127,22 +128,23 @@ def create_arguments(filename, file, config_file,
...
-        if remark_configs_plugins:
-            config_json = json.dumps(remark_configs_plugins)
-            plugins = 'lint=' + config_json[1:-1]
-            args += ['--use', plugins]
+        if remark_lint_configs:
+            config_json = json.dumps(remark_lint_configs)
+            lint = 'lint=' + config_json[1:-1]
+            args += ['--use', lint]
 
+        args += ['--use', 'validate-links']
```
Comment (sils) [inline bears/markdown/MarkdownBear.py]: shouldn't we have an option for this?

File: bears/markdown/MarkdownBear.py [hunk 2]
```diff
@@ -21,15 +21,16 @@ class MarkdownBear:
...
+                    NpmRequirement('remark-validate-links', '5')}
     AUTHORS = {'The coala developers'}
     AUTHORS_EMAILS = {'coala-devel@googlegroups.com'}
     LICENSE = 'AGPL-3.0'
     CAN_FIX = {'Formatting'}
 
     _output_regex = re.compile(
         r'\s*(?P<line>\d+):(?P<column>\d+)'
-        r'\s*(?P<severity>warning)\s*(?P<message>.*)')
+        r'\s*(?P<severity>warning)\s*(?P<message>.+?)(?:  .*|\n|$)')
```
Comment (Makman2) [inline bears/markdown/MarkdownBear.py]: is this maybe a separate commit?

File: bears/markdown/MarkdownBear.py [hunk 3]
```diff
@@ -127,22 +128,23 @@ def create_arguments(filename, file, config_file,
             'ruleSpaces': horizontal_rule_spaces,       # Bool
             'ruleRepetition': horizontal_rule_repeat,   # int
         }
-        remark_configs_plugins = {}
+        remark_lint_configs = {}
```
Comment (Makman2) [inline bears/markdown/MarkdownBear.py]: the rename should be a separate commit

File: bears/markdown/MarkdownBear.py (hunk 1 above)
Comment (Makman2) [inline bears/markdown/MarkdownBear.py]: orientate on our `InvalidLinksBear` regarding setting names :+1:

Comment (yash-nisar) [inline bears/markdown/MarkdownBear.py]: @Makman2 Our `InvalidLinkBear` has no suck setting. I've also checked the `UrlBear` on which the former bear depends. :wink:

File: bears/markdown/MarkdownBear.py (hunk 2 above)
Comment (yash-nisar) [inline bears/markdown/MarkdownBear.py]: Sure :+1:

**Comment by yash-nisar** [comment]
@Makman2 How does it look now ?

File: bears/markdown/MarkdownBear.py (hunk 1 above)
Comment (Makman2) [inline bears/markdown/MarkdownBear.py]: Yeah if you enable `InvalidLinkBear` you definitely want to check invalid links :) In this case this should be optional :+1: Internet accesses aren't always desired by people that want to lint their code :+1:

Comment (yash-nisar) [inline bears/markdown/MarkdownBear.py]: No No, I meant I've checked the `InvalidLinkBear` for a similar setting name. Is the PR good ?

File: bears/markdown/MarkdownBear.py [hunk 4]
```diff
@@ -108,6 +110,9 @@ def create_arguments(filename, file, config_file,
             The number of times the horizontal rule character will be repeated.
         :param max_line_length:
             The maximum line length allowed.
+        :param check_valid_links:
+            Checks if links to headings and files in markdown point to
+            existing things.
```
Comment (Makman2) [inline bears/markdown/MarkdownBear.py]: Hmm that kinda sounds weird :D

File: bears/markdown/MarkdownBear.py [hunk 5]
```diff
@@ -65,7 +66,8 @@ def create_arguments(filename, file, config_file,
                          horizontal_rule: str='*',
                          horizontal_rule_spaces: bool=False,
                          horizontal_rule_repeat: int=3,
-                         max_line_length: int=None):
+                         max_line_length: int=None,
+                         check_valid_links: bool=None):
```
Comment (Makman2) [inline bears/markdown/MarkdownBear.py]: Use `False` instead of `None`, we don't need a tri-state for this field.

File: bears/markdown/MarkdownBear.py [hunk 6]
```diff
@@ -136,13 +141,16 @@ def create_arguments(filename, file, config_file,
         # Remove { and } as remark adds them on its own
         settings = config_json[1:-1]
 
-        args = ['--no-color', '--quiet', '--setting', settings]
+        args = [filename, '--no-color', '--quiet', '--setting', settings]
```
Comment (Makman2) [inline bears/markdown/MarkdownBear.py]: do you really need `filename`, as the linter has `use_stdin`?

File: tests/markdown/MarkdownBearTest.py [hunk 7]
```diff
@@ -50,7 +50,7 @@
 
 
 @generate_skip_decorator(MarkdownBear)
-class MarkdownBearMaxLineLengthMessageTest(unittest.TestCase):
+class MarkdownBearTest(unittest.TestCase):
```
Comment (Makman2) [inline tests/markdown/MarkdownBearTest.py]: for me it's obvious, anyway put the reason into the commit message :)

File: bears/markdown/MarkdownBear.py [hunk 8]
```diff
@@ -30,7 +30,7 @@ class MarkdownBear:
 
     _output_regex = re.compile(
         r'\s*(?P<line>\d+):(?P<column>\d+)'
-        r'\s*(?P<severity>warning)\s*(?P<message>.*)')
+        r'\s*(?P<severity>warning)\s*(?P<message>.+?)(?:  .*|\n|$)')
```
Comment (Makman2) [inline bears/markdown/MarkdownBear.py]: Please reason that inside the commit message :)

File: bears/markdown/MarkdownBear.py [hunk 9]
```diff
@@ -132,21 +132,21 @@ def create_arguments(filename, file, config_file,
...
             'ruleRepetition': horizontal_rule_repeat,   # int
         }
-        remark_configs_plugins = {}
+        remark_lint_configs = {}
 
         if max_line_length:
-            remark_configs_plugins['maximumLineLength'] = max_line_length
+            remark_lint_configs['maximumLineLength'] = max_line_length
 
         config_json = json.dumps(remark_configs_settings)
```
Comment (Makman2) [inline bears/markdown/MarkdownBear.py]: Maybe we should also rename this to `remark_configs`, so it's easier to distinct from `remark_lint_configs`?

Up to you :+1:

File: bears/markdown/MarkdownBear.py (hunk 8 above)
Comment (Makman2) [inline bears/markdown/MarkdownBear.py]: Actually I would rephrase the shortlog to something like `Improve result message` or so, as this is the effective reason :)

**Comment by Asnelchristian** [comment]
please rebase and address @Makman2  issues :)

File: bears/markdown/MarkdownBear.py (hunk 4 above)
Comment (yash-nisar) [inline bears/markdown/MarkdownBear.py]: Changed :+1:

File: bears/markdown/MarkdownBear.py (hunk 5 above)
Comment (yash-nisar) [inline bears/markdown/MarkdownBear.py]: Done :+1:

File: bears/markdown/MarkdownBear.py (hunk 6 above)
Comment (yash-nisar) [inline bears/markdown/MarkdownBear.py]: Yes, `use_stdin` doesn't work with `validate-links` :smile: I've removed the `use_stdin` though :+1:

File: tests/markdown/MarkdownBearTest.py (hunk 7 above)
Comment (yash-nisar) [inline tests/markdown/MarkdownBearTest.py]: Done :+1:

File: bears/markdown/MarkdownBear.py (hunk 8 above)
Comment (yash-nisar) [inline bears/markdown/MarkdownBear.py]: Done :+1:

File: bears/markdown/MarkdownBear.py (hunk 9 above)
Comment (yash-nisar) [inline bears/markdown/MarkdownBear.py]: One more commit ?

Comment (yash-nisar) [inline bears/markdown/MarkdownBear.py]: Done, added a commit stating the same :+1:

**Comment by yash-nisar** [comment]
Updated @Makman2

File: bears/markdown/MarkdownBear.py (hunk 6 above)
Comment (Makman2) [inline bears/markdown/MarkdownBear.py]: Ah write that into the commit message please :D

File: bears/markdown/MarkdownBear.py [hunk 10]
```diff
@@ -65,7 +65,8 @@ def create_arguments(filename, file, config_file,
                          horizontal_rule: str='*',
                          horizontal_rule_spaces: bool=False,
                          horizontal_rule_repeat: int=3,
-                         max_line_length: int=None):
+                         max_line_length: int=None,
+                         check_valid_links: bool=False):
```
Comment (Makman2) [inline bears/markdown/MarkdownBear.py]: maybe just `check_links`? Strictly: As only checking valid links won't yield any results, as those are valid, right? ;)

File: bears/markdown/MarkdownBear.py [hunk 11]
```diff
@@ -130,21 +130,21 @@ def create_arguments(filename, file, config_file,
             'ruleSpaces': horizontal_rule_spaces,       # Bool
```
Comment (Makman2) [inline bears/markdown/MarkdownBear.py]: Please add the reason again into the commit message ;)

**Comment by Makman2** [comment]
For the rename commits: Please put the reason of the name-change into the commit body :+1:

File: bears/markdown/MarkdownBear.py (hunk 6 above)
Comment (yash-nisar) [inline bears/markdown/MarkdownBear.py]: Done :+1:

**Comment by Makman2** [comment]
ack f68bd91 2a21048 d83fe5f 71f6bc5 d6100ce

**Comment by Makman2** [comment]
@rultor merge
